pkg/controllers/managementuser/secret/secret.go

```diff
@@ -256,7 +254,7 @@ func (s *Controller) createOrUpdate(obj *corev1.Secret, action string) error {
 		namespacedSecret := getNamespacedSecret(obj, namespace.Name)
 		switch action {
 		case create:
-			if _, err := s.secretLister.Get(namespacedSecret.Namespace, namespacedSecret.Name); err == nil {
+			if _, err := s.secrets.GetNamespaced(namespacedSecret.Namespace, namespacedSecret.Name, metav1.GetOptions{}); err == nil {
```

Comment:
Is Get-before-Create a valid optimization in general, or will unconditionally calling Create actually perform better?

I am asking because this code was introduced long ago in a PR titled "Performance fixes" which moved from unconditionally Create to checking with a cached Controller first.

[`26f3290` (#35401)](https://github.com/rancher/rancher/pull/35401/commits/26f329055081cc58a9e51dee0761d113cf295709)

@bigkevmcd are there any known mechanisms in which Rancher may create a lot of these Secrets that end up having to be copied to downstream clusters all at the same time?

Reply:
Nice point.
 - Get-before-Create may incur into unnecessary data received "from" downstreams, but prevent unnecessary data being sent "to" the downstream.
 - Create may incur into unnecessary data sent "to" the downstream.
I think both cases are equally expensive, it's just a matter of the direction of the transfer, in which case, I think I'll go with the second option.

I think this is one drawback we would need to live with in order to achieve removing downstream caches for secrets. Unfortunately, the Kubernetes does not allow checking existence or partially retrieving a resource.

The only option I can think of is maybe letting the agent use a cache, then have a way for Rancher to query it?

Reply:
I think we can expect reads to be cheaper than writes in general. Reading a bit around etcd I think I understand that any write operation needs 1) consensus, and that incurs intra-node network latency and 2) flushing, and therefore disk write latency.

Read operations, IIUC, can be served by a single node (best case) and require no flushing.

So while Get-Before-Create may generate more network traffic, I'd expect it to be more efficient from a time and I/O perspective under the assumption a good percentage of Writes can be saved.

Does that make sense?

Reply:
Other than the first reconciliation, reading first will optimise the process.

It does introduce small race windows, but nothing that I'd be particularly worried about, and as @moio points out, it's a lot cheaper to read than write.

Reply:
I can't see any cases where this particular controller would be automatically triggered, but someone from @rancher/collie might know better than me.

Reply:
I agree read first is more optimal. 
This will be triggered only for secrets inside the cluster namespace that have the label `cattle.io/creator=norman`, so I wouldn't expect this to be triggered many times.